### PR TITLE
Mempool refactor

### PR DESCRIPTION
**Pull request by lunfardo314**

### CODE REVIEW

File: packages/chain/mempool/mempool.go [hunk 1]
```diff
@@ -191,15 +191,22 @@ func (m *mempool) traceIn(req iscp.Request) {
 	if traceInOut {
 		logFn = m.log.Infof
 	}
-	tl := time.Time{}
+	var untilTime time.Time
+	var untilMilestone uint32
```
Comment (jorgemmsilva) [inline packages/chain/mempool/mempool.go]: hard to understand what `untilTime` means. Propose renaming to: `timelockTime` and `timelockMilestone`

File: packages/vm/core/accounts/impl.go [hunk 2]
```diff
@@ -113,18 +113,29 @@ func withdraw(ctx iscp.Sandbox) (dict.Dict, error) {
...
+	// It is needed for subsequent Send call
 	a.Require(MoveBetweenAccounts(state, ctx.Caller(), commonaccount.Get(ctx.ChainID()), tokensToWithdraw),
 		"accounts.withdraw.inconsistency. failed to move tokens to owner's account")
-
-	// add incoming tokens (after fees) to the balances to be withdrawn. Otherwise they would end up in the common account
+	// add incoming tokens (after fees) to the balances to be withdrawn.
+	// Otherwise, they would end up in the common account
 	tokensToWithdraw.Add(ctx.IncomingTransfer())
+	// Now all caller's assets are in the common account
+	// TODO: by default should be withdrawn all tokens and account shpoild eb closed
```
Comment (jorgemmsilva) [inline packages/vm/core/accounts/impl.go]: >shpoild eb closed

? :sweat_smile:

**Review by jorgemmsilva — APPROVED** [review]

File: packages/transaction/util.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+	return inputs, remainder, nil
+}
+
+// computeReminderOutput calculates reminders for iotas and native tokens, returns skeleton reminder output
+// which only contains assets filled in.
+// - inIotas and inTokens is what is available in inputs
+// - outIotas, outTokens is what is in outputs, except the reminder output itself with its dust deposit
+// Returns nil if inputs are not enough (taking into account dust deposit requirements)
+// If return not nil but Amount == 0, ite means reminder is a perfect match between inputs and outputs, remainder not needed
+func computeReminderOutput(senderAddress iotago.Address, inIotas, outIotas uint64, inTokens, outTokens map[iotago.NativeTokenID]*big.Int, deSeriParams *iotago.DeSerializationParameters) *iotago.ExtendedOutput {
```
Comment (dessaya) [inline packages/transaction/util.go]: typo: `reminder` -> `remainder`

File: packages/vm/core/governance/interface.go [hunk 4]
```diff
@@ -72,18 +73,36 @@ const (
...
+		ParamValidatorFee,
+		ParamHname,
+		ParamChainID,
+		ParamDescription,
+		ParamMaxBlobSize,
+		ParamMaxEventSize,
+		ParamMaxEventsPerRequest) {
+		panic("wrong constant in governance/interface.go")
+	}
+}
```
Comment (dessaya) [inline packages/vm/core/governance/interface.go]: At this point I think we should be using enums instead? If we want them to be ascii could make them go like a b c ...:

```
ParamChainOwner = kv.Key('a' + iota)
ParamOwnerFee
ParamChainOwner
ParamOwnerFee
...
```

**Review by dessaya — APPROVED** [review]

File: packages/chain/mempool/mempool.go [hunk 5]
```diff
@@ -222,31 +222,40 @@ func (m *mempool) traceOut(reqid iscp.RequestID) {
...
+	expiry, expiryAddress := r.Expiry()
+	if expiryAddress != nil {
+		if !expiry.Time.After(currentTime.Add(FallbackDeadlineMinAllowedInterval)) {
 			return false, true
 		}
 	}
+
 	// time lock
-	return r.TimeLock().IsZero() || r.TimeLock().Before(nowis), false
+	return r.TimeLock().Time.IsZero() || r.TimeLock().Time.Before(currentTime), false
```
Comment (jorgemmsilva) [inline packages/chain/mempool/mempool.go:245]: this should use the common check we talked about initially, I think it's already available in iota.go

Comment (lmoe) [reply packages/chain/mempool/mempool.go:245]: Yup, I was about to put this into the vm validator method as well and use it in both.

**Pull request merged**